### ISSUE TITLE
Feature Request: Substring from StringBuilder

### ISSUE THREAD

**Issue by varunu28** [linkdotnet/golang-stringbuilder#5]
Copy a sub string from StringBuilder to an external data structure such as Array

https://docs.oracle.com/javase/8/docs/api/java/lang/StringBuilder.html#getChars-int-int-char:A-int-

```
sb = sb.Append("a").Append("b").Append("c")
r := make([]rune, 2)
sb.Substring(0,2,r) -> "ab"
```

**Comment by linkdotnet** [comment]
Hey @varunu28,

thanks for your proposals. In general I created the stringbuilder based on my "zero-allocation" StringBuilder in C#: https://github.com/linkdotnet/StringBuilder

A question to the public API: Is there a reason to put the substring into `r` instead of returning the whole thing?

**Comment by varunu28** [comment]
> A question to the public API: Is there a reason to put the substring into r instead of returning the whole thing?

If we are trying to reuse an existing data structure either by overwriting on top of it or just re-using previously allocated space then, it is better to provide the source to write to instead of moving that responsibility of space allocation to `StringBuilder` class. In Java constructs like `Arrays.copy` also follow this behavior where they take the original array and the array to which we want to copy as input. 

Essentially the `StringBuilder` class should not be managing any memory which it is not going to directly use and instead use the resources provides by the client to perform mutation.

**Comment by linkdotnet** [comment]
That is an interesting thought. The one in Java and C# does indeed return a new object.

But if we go down that path, it makes sense that the method will return at least an `error` in the case of:
 * The slice isn't big enough
 * Start / End is "invalid" - like out of bounds or start is after end

**Comment by varunu28** [comment]
> That is an interesting thought. The one in Java and C# does indeed return a new object.
> 
> But if we go down that path, it makes sense that the method will return at least an `error` in the case of:
> 
> * The slice isn't big enough
> * Start / End is "invalid" - like out of bounds or start is after end

Definitely, we will need to have error handling for above mentioned scenarios

**Comment by linkdotnet** [comment]
The more I think about it, I still feel the StringBuilder should be responsible for creating the slice and returning it, instead of getting it passed in.
It would definitely reduce the surface area of bugs for users - and our code is much simpler.

So the function would look like this:
```golang
func (s *StringBuilder) Substring(start int, end int) (string, error)
```

On top we already have a `AsRuneSlice` method which does a similar thing, so maybe `Slice` might be a better name.

The idiomatic way in golang to get a subslice would be:
`s = s[0:len(s)-3]`. It might feel to java-ish/C#-ish if we have a `Substring(inclStart int, exclEnd int)` method. But these are just my thoughts on top of my head.

**Comment by varunu28** [comment]
> The more I think about it, I still feel the StringBuilder should be responsible for creating the slice and returning it, instead of getting it passed in. It would definitely reduce the surface area of bugs for users - and our code is much simpler.
> 
> So the function would look like this:
> 
> ```go
> func (s *StringBuilder) Substring(start int, end int) (string, error)
> ```
> 
> On top we already have a `AsRuneSlice` method which does a similar thing, so maybe `Slice` might be a better name.
> 
> The idiomatic way in golang to get a subslice would be: `s = s[0:len(s)-3]`. It might feel to java-ish/C#-ish if we have a `Substring(inclStart int, exclEnd int)` method. But these are just my thoughts on top of my head.

Ok now that I see the implementation of both `AsRuneSlice` and `AsRuneArray`, I feel this shouldn't have been implemented this way. With existing implementation the following scenario is possible:
```
sb := NewStringBuilderFromString("hello")
r := sb.AsRuneSlice()
r[0] = 't'

fmt.Println(sb.ToString()) // prints "tello"
```

This behavior seems like a side-effect and kind of breaks abstraction as we allow to mutate the internal state of string builder data structure from outside. Any form of mutation should happen internally and the data structure should have complete control over it. Very basic scenario where this will end up failing is say if we want to log(or keep a counter) every time a string builder is updated. In the above sample code, we fail to do so as we don't have complete control over mutation. 

LMK if this makes sense. I feel the right behavior should be to not share the internal data representation with the client and instead share a copy of it in a similar way we do for `ToString` method.

**Comment by linkdotnet** [comment]
This are very valid concerns - and maybe the docs are not enough at the moment (they state this is the internal slice you get returned).
To be fair, I leaned towards the C# implementation that allows this kind of mutation, as you can get the underlying memory.

For the time being we could also use the following approach:
```golang
func (s *StringBuilder) AsRuneSlice() []rune {
	r := make([]rune, s.position)
	copy(r, s.data[:s.position])
	return r
}
``` 

But I can see that this isn't ideal as well. The good thing is that I wanted to go to `v1` anyway soon - this could be something where we break the current API surface to something we can live with.

The way described above would also work for `Substring` as we copy the contents into a new slice rather than returning the underlying memory-pointer.

My way of thinking was that the API surface now is very user-friendly. You don't have to take care of anything. The user would always have this pattern:

```golang
r := make([]rune, 10)
sb.Substring(0, 10, r);
```

Basically disallows some errors. And I did see this as the major (>80%) use case. Your proposed alternative would allow the user to slice their own slice to add the substring somewhere in the middle of his slice. I figured that might be somewhat an edge case. But then again I am open to change that. 

I would just like to have a stable API for `v1`.

**Comment by varunu28** [comment]
Definitely. I agree that as these are changes in the API contract, they should be part of the `v1` API. Another requirement that we can start thinking of is that do we want to consider thread-safety of string builder in `v1` API. With the existing implementation, we don't have thread-safety guarantee. 

I propose starting an epic where we can list down the roadmap for `v1` API and then tackle these ideas as part of design discussion once we have clarity about what are we planning to deliver in the `v1`.

**Comment by varunu28** [comment]
Also we can push the substring functionality to `v1` instead of having a separate behavior that differs from existing implementation.

**Comment by linkdotnet** [comment]
Oh don't get me wrong, we can totally introduce that in `v1` - your request just made it visible that we have some inconsistencies in the current approach, which we should correct in the next major version.

So please feel free to implement this with your approach. I do understand your reasoning and just because I brought up Java and C# doesn't mean we should do it the same way here. Sorry if that was transported that way.

If we both decide it would be good to have buffer passed towards the builder rather than creating it on our own, I do think `Substring` is a good example to implement right now

**Comment by varunu28** [comment]
Resolved under https://github.com/linkdotnet/golang-stringbuilder/pull/8
